Add imports to surface area

dacamo76 opened the pull request:
Currently only adds aliased imports to the surface area.

- [x] Add aliased symbols to surface area: `import numpy as np`
- [x] Add aliased from imports: `from numpy import ones as l`
- [x] Add non-aliased from imports: `from numpy import ones`
- [x] Add non-aliased imports: `import numpy`
- [ ] Add star imports??? not sure if this is needed since we don't know imported symbols

CJ-Wright commented:
For star imports, my current thinking is that we leave something in the way of metadata so that if a 3rd party cares about them they can go chase them down. What do you think?

dacamo76 commented:
> For star imports, my current thinking is that we leave something in the way of metadata so that if a 3rd party cares about them they can go chase them down. What do you think?

yes, but not sure how to express that in the data model.

For now I added a `type` of "import" and added an attribute called "shadows" which lists the symbol it is shadowing.

The following code:
```python
import numpy as np
from abc import xyz as l
from ggg import efg
import ghi

z = np.ones(5)
```

is represented by:
```python
z.symbols == {
        "mm.np": {"shadows": "numpy", "lineno": None, "type": "import"},
        "mm.l": {"shadows": "abc.xyz", "lineno": None, "type": "import"},
        "mm.efg": {"shadows": "ggg.efg", "lineno": None, "type": "import"},
        "mm.ghi": {"shadows": "ghi", "lineno": None, "type": "import"},
        "mm": {"lineno": None, "symbols_in_volume": {"numpy.ones"}, "type": "module"},
        "mm.z": {"lineno": 7, "symbols_in_volume": set(), "type": "constant"},
    }
```

This works for non-star imports since we know the specific symbol being shadowed, how do we want to represent star imports?

something like this maybe?
```python
from numpy import *
from abc import *

z.symbols == {
        "*": {"star_imported": ["abc", "numpy"], "lineno": None, "type": "import"}
}
```
where the attribute "*" is special?

dacamo76 commented:
@CJ-Wright I think this PR  brings enough value that we should separate it from the star import question. We can move the discussion of the star import data model to that new issues to hash it out a bit more.

dacamo76 commented on tests/test_ast_symbol_extractor.py:
```diff
@@ -128,31 +153,29 @@ def xyz(self):
     z = SymbolFinder(module_name="mm")
     z.visit(tree)
-    # TODO: should we add a key in the metadata to say that a symbol is
-    #  a reference to another symbol?
```
I think the type "import" satisfies this requirement.

dacamo76 commented on tests/test_ast_symbol_extractor.py:
```diff
@@ -128,31 +153,29 @@ def xyz(self):
-        "mm.l": {},
-        'mm': {'lineno': None, 'symbols_in_volume': set(), 'type': 'module'},
-        "mm.z": {"lineno": 4, "symbols_in_volume": {"numpy.ones"}, "type": "constant"},
+        "mm.np": {"shadows": "numpy", "lineno": None, "type": "import"},
```
@CJ-Wright would like your comments on this data model.
Do we want all values of symbols to have the same attributes?
The way it currently stands the `type` determines the attributes in the object, we might consider lifting it out of the values and into a first-class citizen.

Something like this keeps it easy to iterate over all while also simple to filter out unneeded types, would also help keeping all the types data model consistent with each other.
Also may help expose/consume later on via automated tools/schemas. 😄 
```python
z.symbols == {
        "mm.np": {
                "type": "import",
                "data": {"shadows": "numpy", "lineno": None,},
        },
        "mm.f": {
                "type": "function",
                "data": {"symbols_in_volume": set(), "lineno": 5}
       }
```

CJ-Wright replied:
That looks great! It most likely also helps with less flexable json consumers that expect the keys to be consistent in the data model (or at least consistent to a particular type in this case).

CJ-Wright reviewed (approved):
LGTM, do you want me to wait for the new data model before merging?

dacamo76 commented:
I created a new issue to hash out the data model.
Will merge this.
